man/roverlay.1
.\" groff -Tascii -man
.TH "ROVERLAY" "1" "July 9 2012" "Linux/Gentoo?" "R Overlay"

.SH "NAME"
roverlay \- create and maintain an overlay of R packages

.SH "SYNOPSIS"
.TP
.BR "roverlay " "[\fIoptions\fR] [\fIcommand\fR]..."
.TP
.BR "roverlay " "\fB\-h\fR | \fB\-\-help\fR"
.TP
.BR "roverlay " "\fB\-\-help-config\fR"
.TP
.BR "roverlay " "\fB\-\-print-config\fR"
.TP
.BR "roverlay " "\fB\-V\fR | \fB\-\-version\fR"
.SH "DESCRIPTION"
todo
.SH "COMMANDS"
.TP
.BR sync
Sync all repos. Has no effect if you don't specify at least one remote repository.
.TP
.BR create
Create the overlay - read R packages and try to create ebuilds for them.
Implies \fBsync\fR unless \fB\-\-nosync\fR is given.
This is the default command.
.TP
.BR nop
no-op. To be removed.
.SH "OPTIONS"
.TP
.BR "\-\-help " "(\fB\-h\fR)"
Print a help message and exit
.TP
.BR "\-\-version " "(\fB\-V\fR)"
Print the program's version and exit
.TP
.BR "\-\-help\-config"
Print all known config options and exit
.TP
.BR "\-\-print\-config"
Print the config tree after reading the config file and exit
.SS "configuration files"
.TP
.BR "\-\-config " "\fIfilename\fR"
Read the main configuration from \fIfilename\fR instead of ${PWD}/R-overlay.conf
.TP
.BR "\-\-deprule\-file " "\fIfilename\fR (\fB\-D\fR)"
Read dependency rules from \fIfilename\fR instead of reading the
files listed in the main config (\fISIMPLE_RULES_FILE\fR).
Can be specified more than once.
.TP
.BR "\-\-field\-definition " "\fIfilename\fR (\fB\-F\fR)"
Read the field definition from \fIfilename\fR instead
of the \fIFIELD_DEFINITION\fR file specified in the main config.
.TP
.BR "\-\-repo\-config " "\fIfilename\fR (\fB\-R\fR)"
Read the repo config from \fIfilename\fR instead
of the files listed in the main config (\fIREPO_CONFIG\fR).
Can be specified more than once.
.SS "R packages"
.TP
.BR "\-\-distdir" "\fR, " "\-\-from " "\fIdirectory\fR"
Create ebuilds for packages in \fIdirectory\fR.
Implies \fBcreate\fR and \fB\-\-nosync\fR.
Disables all other repos.
.TP
.BR "\-\-distroot " "\fIdirectory\fR"
Set the distroot to \fIdirectory\fR.
This is used as root for all per-repo package dirs that don't define another location.
.TP
.BR "\-\-force\-distroot"
Use the distroot even for repos that define their own location.
.TP
.BR "\-\-nosync" "\fR, " "\-\-no\-sync"
Disable repo syncing (offline mode). Blocks the \fBsync\fR command.
.SS "overlay"
.TP
.BR "\-\-show" "\fR, " "\-\-show\-overlay"
Print ebuilds/metadata after creation to stderr.
.TP
.BR "\-\-write" "\fR, " "\-\-write\-overlay"
Write the whole overlay after creation.
This includes ebuilds, metadata, Manifest files and the profiles/ dir.
.TP
.BR "\-\-stats" "\fR, " "\-\-no\-stats"
Print some stats after overlay creation/writing.
.TP
.BR "\-\-no\-manifest"
Skip Manifest file creation to save time. Note that the resulting overlay
cannot be used with portage/emerge.
.TP
.BR "\-\-overlay " "\fIdirectory\fR (\fB\-O\fR)"
Set the overlay directory to \fIdirectory\fR. Else
\fIOVERLAY_DIR\fR from the main config will be used.
Implies \fB\-\-write\fR.
.TP
.BR "\-\-overlay\-name " "\fIname\fR (\fB\-N\fR)"
Set the overlay name to \fIname\fR.
Else \fIOVERLAY_NAME\fR from the main config will be used.
.SH "EXAMPLES"
.TP
roverlay \-c \fIconfig\fR
Run overlay creation with sync, all options (overlay name/dir,...) are read from \fIconfig\fR
.TP
roverlay \-\-from \fIpkgdir\fR \-O \fIoverlay\fR \-N R_Overlay \-\-field-definition \fIfield_definition\fR \-D \fIdeprules\fR
Run overlay creation for packages from \fIpkgdir\fR and write the overlay named R_Overlay to \fIoverlay\fR.
Use the given \fIfield_definition\fR and \fIdeprules\fR files.
.br
!!! This should do without a config file (FIXME)
.SH "SEE ALSO"
.BR "roverlay\-repo(5)",
.BR "roverlay\-deprules(5)"


roverlay.py
#!/usr/bin/python -OO
# -*- coding: utf-8 -*-
# R overlay -- main script
# Copyright (C) 2012 André Erdmann <dywi@mailerd.de>
# Distributed under the terms of the GNU General Public License;
# either version 2 of the License, or (at your option) any later version.

"""main script for R overlay creation"""

__all__ = [ 'DIE', 'roverlay_main' ]

import os
import sys

# roverlay modules will be imported later

HIDE_EXCEPTIONS = False

class DIE ( object ):
	"""Container class for various system exit 'events'."""
	NOP          =  os.EX_OK
	ERR          =  1
	BAD_USAGE    =  os.EX_USAGE
	USAGE        =  os.EX_USAGE
	ARG          =  9
	CONFIG       =  os.EX_CONFIG
	OV_CREATE    =  20
	SYNC         =  30
	CMD_LEFTOVER =  90
	IMPORT       =  91
	UNKNOWN      =  95
	INTERRUPT    = 130

	@staticmethod
	def die ( msg=None, code=None ):
		"""
		Calls syst.exit (code:=DIE.ERR) after printing a message (if any).
		"""
		code = DIE.ERR if code is None else code
		if msg is not None:
			sys.stderr.write ( msg + "\n" )
#		else:
#			sys.stderr.write ( "died.\n" )
		sys.exit ( code )
	# --- end of die (...) ---

# --- DIE: exit codes ---
die = DIE.die

def roverlay_main():
	"""roverlay.py main() - parse args, run overlay creation, sync, ..."""
	def optionally ( call, option, *args, **kw ):
		if OPTION ( option ):
			return call ( *args, **kw )
	# --- end of optionally (...) ---

	#repo_list = None
	#overlay   = None
	def run_sync():
		if "sync" in actions_done: return
		try:
			# set up the repo list
			global repo_list
			repo_list = RepoList (
				sync_enabled   = not OPTION ( 'nosync' ),
				force_distroot = OPTION ( 'force_distroot' )
			)

			## extra_opts->distdir ... TODO
			if 'distdirs' in extra_opts:
				repo_list.add_distdirs ( OPTION ( 'distdirs' ) )
			else:
				# default repo list
				repo_list.load()

			## this runs _nosync() or _sync(), depending on extra_opts->nosync
			repo_list.sync()

			set_action_done ( "sync" )

		except KeyboardInterrupt:
			die ( "Interrupted", DIE.INTERRUPT )
		except:
			if HIDE_EXCEPTIONS:
					die (
						"nosync() failed!" if OPTION ( "nosync" ) \
							else "sync() failed!",
						DIE.SYNC
					)
			else:
				raise
	# --- end of run_sync() ---

	def run_overlay_create():
		if "create" in actions_done: return
		#run_sync()
		try:
			global overlay
			overlay = OverlayCreator (
				skip_manifest=OPTION ( 'skip_manifest' ),
				allow_write=OPTION ( 'write_overlay' )
			)

			repo_list.add_packages ( overlay.add_package )

			overlay.run ( close_when_done=True )

			optionally ( overlay.write_overlay, 'write_overlay' )
			optionally ( overlay.show_overlay,  'show_overlay'  )
			if OPTION ( 'print_stats' ): print ( "\n" + overlay.stats_str() )

			set_action_done ( "create" )

		except KeyboardInterrupt:
			die ( "Interrupted", DIE.INTERRUPT )
		except:
			if HIDE_EXCEPTIONS:
				die ( "Overlay creation failed.", DIE.OV_CREATE )
			else:
				raise
		finally:
			if 'overlay' in locals() and not overlay.closed:
				# This is important 'cause it unblocks remaining ebuild creation
				# jobs/threads, specifically waiting EbuildJobChannels in depres.
				# It also writes the deps_unresolved file
				overlay.close()
	# --- end of run_overlay_create() ---

	# get args
	# imports roverlay.argutil (deleted when done)
	try:
		import roverlay.argutil
	except ImportError:
		if HIDE_EXCEPTIONS:
			die ( "Cannot import roverlay modules!", DIE.IMPORT )
		else:
			raise

	COMMAND_DESCRIPTION = {
		'sync'           : 'sync repos',
		'create'         : 'create the overlay '
								  '(implies sync, override with --nosync)',
		'depres_console' : \
			'run an interactive depres console (highly experimental)',
		'depres'         : 'this is an alias to \'depres_console\'',
		'nop'            : 'does nothing',
	}

	commands, config_file, additional_config, extra_opts = \
		roverlay.argutil.parse_argv (
			command_map=COMMAND_DESCRIPTION,
			default_config_file="R-overlay.conf"
		)

	OPTION = extra_opts.get

	del roverlay.argutil

	# -- load config

	# imports: roverlay, roverlay.config.entryutil (if --help-config)

	try:
		import roverlay
	except ImportError:
		if HIDE_EXCEPTIONS:
			die ( "Cannot import roverlay modules!", DIE.IMPORT )
		else:
			raise

	try:
		roverlay.setup_initial_logger()

		conf = roverlay.load_config_file (
			config_file,
			extraconf=additional_config
		)
		del config_file, additional_config
	except:
		if HIDE_EXCEPTIONS:
			die (
				"Cannot load config file {!r}.".format ( config_file ), DIE.CONFIG
			)
		else:
			raise

	if OPTION ( 'list_config' ):
		try:
			from roverlay.config.entryutil import list_entries
			print ( "== main config file ==\n" )
			print ( list_entries() )
		except:
			raise
			die ( "Cannot list config entries!" )

		EXIT_AFTER_CONFIG = True

	if OPTION ( 'print_config' ):
		try:
			conf.visualize ( into=sys.stdout )
		except:
			die ( "Cannot print config!" )
		EXIT_AFTER_CONFIG = True


	if 'EXIT_AFTER_CONFIG' in locals() and EXIT_AFTER_CONFIG:
		sys.exit ( os.EX_OK )


	# -- determine commands to run
	# (TODO) could replace this section when adding more actions
	# imports roverlay.remote, roverlay.overlay.creator

	actions = set ( filter ( lambda x : x != 'nop', commands ) )

	if 'sync' in actions and OPTION ( 'nosync' ):
		die ( "sync command blocked by --nosync opt.", DIE.ARG )

	del commands


	if not actions:
		# this happens if a command is nop
		die ( "Nothing to do!", DIE.NOP )

	if 'depres_console' in actions or 'depres' in actions:
		if len ( actions ) != 1:
			die ( "depres_console cannot be run with other commands!", DIE.USAGE )

		try:
			from roverlay.depres.simpledeprule.console import DepResConsole
			con = DepResConsole()
			con.run()
			sys.exit ( os.EX_OK )
		except ImportError:
			if HIDE_EXCEPTIONS:
				die ( "Cannot import depres console!", DIE.IMPORT )
			else:
				raise
		except:
			if HIDE_EXCEPTIONS:
				die ( "Exiting on console error!", DIE.ERR )
			else:
				raise



	# -- import roverlay modules

	try:
		from roverlay.remote          import RepoList
		from roverlay.overlay.creator import OverlayCreator
	except ImportError:
		if HIDE_EXCEPTIONS:
			die ( "Cannot import roverlay modules!", DIE.IMPORT )
		else:
			raise

	# -- run methods (and some vars)
	# imports: nothing

	actions_done = set()
	set_action_done = actions_done.add

	# -- run

	# always run sync 'cause commands = {create,sync}
	# and create implies (no)sync
	run_sync()

	if 'create' in actions: run_overlay_create()

	if len ( actions ) > len ( actions_done ):
		die (
			"Some actions (out of {!r}) could not be performed!".format (
				actions ), DIE.CMD_LEFTOVER
		)
# --- end of main() ---

if __name__ == '__main__':
	roverlay_main()
elif not 'pydoc' in sys.modules:
	die ( "Please don't import this script...", DIE.BAD_USAGE )
